lib/Sema/TypeCheckType.cpp

```diff
@@ -1706,7 +1706,8 @@ static bool validateAutoClosureAttributeUse(DiagnosticEngine &Diags,
     return false;
 
   // If is a parameter declaration marked as @autoclosure.
-  if (options.is(TypeResolverContext::FunctionInput)) {
+  if (options.is(TypeResolverContext::FunctionInput) ||
+      options.is(TypeResolverContext::VariadicFunctionInput)) {
```

Comment:
In your example the type is `String`. Should the parser be fixed not to produce `VariadicFunctionInput` for cases like that and diagnose inline?

Reply:
It's set by `resolveASTFunctionTypeParams` when resolving the types. I am not quite sure how we could fix this in the parser.

Reply:
The problem is that presence of `@autoclosure` here makes the type to be `() -> String...` implicitly which means that the parameter is actually still `FunctionInput` with variadic result which is invalid. In `resolveASTFunctionTypeParams` you can check for presence of `@autoclosure` attribute before checking for `hasEllipsis` and integrate new information into that check, plus produce a custom diagnostic for it in `validateAutoClosureAttr`.

Reply:
I see, the missing bit was the implicit conversion. Thank you for the explanation! I'll update the check.

Reply:
No problem! Yeah, the type declared as a parameter marked as `@autoclosure` is actually a result type of implicitly generated closure

Reply:
Oops, it's actually set by `validateParameterType`. It's a bit tricky to check if the result type is variadic, because we just get type in `validateAutoClosureAttr`. Does it make sense to introduce a new `TypeResolverContext` for it specifically?

Reply:
Since `validateParameterType` gets a `ParamDecl` you can check `isAutoClosure()` on it to determine what is going on, same way as in `resolveASTFunctionTypeParams` because `@autoclosure` attribute applies to the declaration of the parameter instead of its type. 

I consider this less of a diagnostic and more of a correctness problem, because having `@autoclosure` flag on the parameter is higher priority that it being variadic because of the implicit function type.

Reply:
Yeah, that's exactly what I have done. I was talking about the custom diagnostic you mentioned above, because then `validateAutoClosureAttr` would need to know that it's variadic and so we would need some way to propagate that information.

Anyway, if having a custom diagnostic isn't really needed then let's just fix the crash.

Reply:
I wonder if should just diagnose that problem inline in both cases then?

Reply:
I guess so. I was thinking to turn this into a request so we can have all the logic for checking an @autoclosure parameter in a single place. What do you think? We can kick off the request in `validateParameterType` or somewhere appropriate.

Reply:
That sounds good to me! But I'm not an expert on requests :) /cc @CodaFi would have better insight about that.

Reply:
I think a more pressing request would be its caller `TypeResolution::resolveType`.  But it’s not a bad idea!

Reply:
Alright then, I'd say we keep this minimal and cherry-pick to the 5.2 branch as well.

Reply:
Sounds good, for now I have created a request to check the "structure" of `@autoclosure`.